### ISSUE TITLE
Simple set up - how to use?

### ISSUE THREAD

**Issue by duggster** [EvanBacon/expo-apple-targets#23]
Hi! I am really looking forward to incorporating a widget into my expo app. In order to get to know this plugin, I am first trying the simplest thing I can think of but I am not able to get it to work. Here are the steps I'm doing, maybe someone can help point me to what I'm doing wrong or at least give some tips on how to troubleshoot. I've been working with expo for a number of years but do not have much (any) experience working in Xcode. 

**Create new app**
1. npx create-expo-app my-app
2. npm run iOS
3. Verify the app opens in the iOS simulator with Expo Go

**Prebuild and add widget in Xcode**
1. npx expo prebuild
2. Set the bundle identifier as com.duggster.myapp
3. Open ./ios/myapp.xcworkspace in Xcode
4. Click File -> New -> Target… -> Widget Extension
- Product Name: my-app-widget
- Dev Team: select my team
- Keep everything else as default
- Click Finish
- A prompt comes up asking: Activate “my-app-widgetExtension scheme”? This scheme has been created for the “null” target… 
- Click Activate
5. Open my_app_widget.swift in Xcode and see the preview show the widget
6. npx expo run:ios
7. The new dev-client opens in the simulator and I’m able to add the widget as expected

**Install and configure the config plugin**
1. npx expo install @bacons/apple-targets
2. Create a folder in the expo project: ./targets
3. Add expo-target.config.json to ./targets
```
{
    "type": "widget",
    "name": "my-app-widget",
    "colors": {
        "$accent": {
            "color": "red",
            "darkColor": "blue"
        }
    },
    "icon": "../assets/icon.png",
    "frameworks": [
        "SwiftUI"
    ],
    "deploymentTarget": "13.4"
}
```
5. Copy my_app_widget.swift and my_app_widgetBundle.swift from ./ios/my-app-widget to ./targets
6. Add teamId to plugins in app.config.js
```
"plugins": [
      [
        "@bacons/apple-targets",
        {
          "teamId": "my team ID here"
        }
      ]
    ]
```
7. npx expo prebuild —clean
8. npx expo run:ios
9. When the dev-client app loads into the simulator, there is no widget available to add for it.
10. I don't see any mention of the widget in the terminal during the prebuild or run:ios commands, and there is no widget target in the new ios workspace.

Any help would be much appreciated!

**Comment by duggster** [comment]
OK, so, I got it working. Here's what I had to do:

- When adding a ./targets folder in the expo project, you have to create a subfolder for the type of extension you're making. In my case I used ./targets/widget and then added the expo-target.config.json file and .swift files in there.
- The app.json "plugins" field should be "appleTeamId" instead of "teamId"
- In the expo-target.config.json, I had to up the deploymentTarget to 17.0 since the default widget code that Xcode gave me had something in it specific to 17.0

After that I was able to have npx expo run:ios load with available widgets, and can continuously regenerate the prebuild as needed whenever I need to modify the widgets. I assume my workflow will be to work in Xcode using the preview feature to develop the widgets, then copy and paste the swift files into targets/widget when I'm done, and I'll remove the ios folder when done. Hoping EAS Build will work as expected. I still have a long way to go, including working in sharing data between the app and widget, but at least this works for being able to embed a widget into my expo app while remaining in the managed workflow!

**Comment by lourxd** [comment]
Has anyone tried to create a Notification Service Extension? Is it possible with this package?

**Comment by shanoysinc** [comment]
> > Did You figure out how to share data between app and widget?
> 
> If anyone is struggling with the same problem with sharing data, just ensure that You followed above steps and have set "entitlements": { "com.apple.security.application-groups": ["group.BUNDLE.widget"] },
> 
> to yours expo-target.config.json

do we need to add .widget at the end for it work?

**Comment by duggster** [comment]
No, you don't need ".widget". For example, I am using something like "group.com.mycompany.myapp". 

I am using react-native-default-preference on the RN side as a wrapper for Swift's UserDefaults (https://developer.apple.com/documentation/foundation/userdefaults). I have only been concerned with iOS, but assume it would work with Android's SharedPreferences there the same way. There are other libraries out there that do similar things.

In react native (only works with a dev client):
```
//Import the package if not in Expo Go
let DefaultPreference;
if (Constants.executionEnvironment === ExecutionEnvironment.StoreClient) { //use dummy functionality if in Expo Go 
  DefaultPreference = {};
  console.log("DEFAULT PREFERENCE", "dummy");
}
else {
  DefaultPreference = require("react-native-default-preference").default;
  DefaultPreference?.setName(Config.APP_GROUP_NAME); //group.com.mycompany.myapp
  console.log("DEFAULT PREFERENCE", Config.APP_GROUP_NAME);
}

//value must be a string (otherwise use JSON serialization)
export const setDefaultPreference = async (key, value) => {
  console.log("setDefaultPreference", key, value);
  return DefaultPreference?.set(key, "" + value)?.then(() => {
    console.log("setDefaultPreference saved", key, value);
  }
  ).catch((e) => {
    console.log("setDefaultPreference Error", e);
  });
}

export const getDefaultPreference = async (key) => {
  return DefaultPreference?.get(key)?.then((value) => {
    console.log("getDefaultPreference", key, value);
    return value;
  }
  ).catch((e) => {
    console.log("getDefaultPreference Error", e);
  });
}
```

On the Swift side, in widgets.swift:
```
let APP_GROUP_NAME = "group.com.mycompany.myapp"

struct SimpleEntry: TimelineEntry {
...
  func getTimeline(in context: Context, completion: @escaping (Timeline<Entry>) -> ()) {
  ...
  let defaults = UserDefaults(suiteName: APP_GROUP_NAME)?.dictionaryRepresentation()
  let userId = defaults?["userId"] as? String
  //Call web service to load user's data
  ...
```

In the RN app, during my app launch or login routine I call setDefaultPreference("userId", userId); 
This will make the value available to the widget, but the widget timeline must be reloaded by the operating system before the Swift code is run again to access the userId and pull in new data to show in the widget.

**Comment by shanoysinc** [comment]
> No, you don't need ".widget". For example, I am using something like "group.com.mycompany.myapp".
> 
> I am using react-native-default-preference on the RN side as a wrapper for Swift's UserDefaults (https://developer.apple.com/documentation/foundation/userdefaults). I have only been concerned with iOS, but assume it would work with Android's SharedPreferences there the same way. There are other libraries out there that do similar things.
> 
> In react native (only works with a dev client):
> 
> ```
> //Import the package if not in Expo Go
> let DefaultPreference;
> if (Constants.executionEnvironment === ExecutionEnvironment.StoreClient) { //use dummy functionality if in Expo Go 
>   DefaultPreference = {};
>   console.log("DEFAULT PREFERENCE", "dummy");
> }
> else {
>   DefaultPreference = require("react-native-default-preference").default;
>   DefaultPreference?.setName("group.com.mycompany.myapp");
>   console.log("DEFAULT PREFERENCE", "real");
> }
> 
> //value must be a string (otherwise use JSON serialization)
> export const setDefaultPreference = async (key, value) => {
>   console.log("setDefaultPreference", key, value);
>   return DefaultPreference?.set(key, "" + value)?.then(() => {
>     console.log("setDefaultPreference saved", key, value);
>   }
>   ).catch((e) => {
>     console.log("setDefaultPreference Error", e);
>   });
> }
> 
> export const getDefaultPreference = async (key) => {
>   return DefaultPreference?.get(key)?.then((value) => {
>     console.log("getDefaultPreference", key, value);
>     return value;
>   }
>   ).catch((e) => {
>     console.log("getDefaultPreference Error", e);
>   });
> }
> ```
> 
> On the Swift side, in widgets.swift:
> 
> ```
> let APP_GROUP_NAME = "group.com.mycompany.myapp"
> 
> struct SimpleEntry: TimelineEntry {
> ...
>   func getTimeline(in context: Context, completion: @escaping (Timeline<Entry>) -> ()) {
>   ...
>   let defaults = UserDefaults(suiteName: APP_GROUP_NAME)?.dictionaryRepresentation()
>   let userId = defaults?["userId"] as? String
>   //Call web service to load user's data
>   ...
> ```
> 
> In the RN app, during my app launch or login routine I call setDefaultPreference("userId", userId); This will make the value available to the widget, but the widget timeline must be reloaded by the operating system before the Swift code is run again to access the userId and pull in new data to show in the widget.

whats the best way to test that data is coming from the app to the widget ?

**Comment by shanoysinc** [comment]
Its working now, I was parsing the json data wrong in swift

**Comment by EvanBacon** [comment]
There's now a native module for sharing data with NSUserDefaults and a CLI to bootstrap targets.
